[TECH] Utiliser le nouveau service d'authentification OIDC pour la CNAV et Pôle Emploi (PIX-5388).

AndreiaPena opened the pull request:
## :unicorn: Problème
Il y a de plus en plus de demande d'interconnexion SSO mais actuellement nous n'avons pas de brique générique et le code est dupliqué à chaque nouvelle demande.


## :robot: Solution
Dans le but de créer la brique générique SSO côté API, nous proposons ici de mutualiser les deux services partenaire en un service d'authentification OIDC

C'est à dire que :
Ces fichiers là disparaîssent : 
- `api/lib/domain/services/authentication/cnav-authentication-service.js`
- `api/lib/domain/services/authentication/pole-emploi-authentication-service.js`

Au profit de : 
- `api/lib/domain/services/authentication/oidc-authentication-service.js` 
- `api/lib/domain/services/authentication/pole-emploi-oidc-authentication-service.js`
- `api/lib/domain/services/authentication/cnav-oidc-authentication-service.js`

La suite de la mutualisation oidc côté API se fera dans une autre PR. 

## :rainbow: Remarques
Pour Pole Emploi nous utilisons le `sub` plutôt que `idIentiteExterne` afin d'être plus dans les standards OIDC.
En effet, d'après [la doc API de Pôle Emploi](https://peio.pe-qvr.fr/data/api/pole-emploi-connect?tabgroup-api=documentation&doc-section=api-doc-section-peconnect), l'idToken contient aussi le `sub`, identique à l'`idIentiteExterne` que nous utilisons actuellement. 
<img width="470" alt="Capture d’écran 2022-07-22 à 16 49 12" src="https://user-images.githubusercontent.com/58915422/180465432-230767f3-10b6-481c-b591-57fb8120ae69.png">

## :100: Pour tester
Tester la non régression sur les connexion Pole Emploi et CNAV

AndreiaPena committed:
feat(api): use oidc authentication service new function in pole emploi and cnav usecase

AndreiaPena committed:
feat(api): add parameters in cnav oidc authentication services

AndreiaPena committed:
recycle(api): remove unused functions

AndreiaPena committed:
recycle(api): remove unused functions

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr4692.review.pix.fr
- App (.org): https://app-pr4692.review.pix.org
- Orga: https://orga-pr4692.review.pix.fr
- Certif: https://certif-pr4692.review.pix.fr
- Admin: https://admin-pr4692.review.pix.fr
- API: https://api-pr4692.review.pix.fr/api/

 Please check it out!

AndreiaPena committed:
feat(api): add getUserInfo in oidc authentication service

AndreiaPena committed:
feat(api): use oidc authentication service new function in pole emploi and cnav usecases

AndreiaPena committed:
api: replace idIdentiteExterne with sub

In fact idIdentiteExterne is a sub alias.
So using idIdentiteExterne is not required and we can use sub for our
standards.

See : https://peio.pe-qvr.fr/data/api/pole-emploi-connect?tabgroup-api=documentation&doc-section=api-doc-section-peconnect

Anne-Gaelle-S commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -19,6 +26,34 @@ class OidcAuthenticationService {
+      throw new AuthenticationTokenRetrievalError(errorMessage, response.code);
+    }
+
+    return new AuthenticationSessionContent({
```
plutôt `return response.data['id_token'];`  ?

Anne-Gaelle-S commented on api/lib/domain/usecases/authentication/authenticate-cnav-user.js:
```diff
@@ -16,7 +16,7 @@ module.exports = async function authenticateCnavUser({
     logger.error(`State sent ${stateSent} did not match the state received ${stateReceived}`);
     throw new UnexpectedOidcStateError();
   }
-  const idToken = await cnavAuthenticationService.exchangeCodeForIdToken({ code, redirectUri });
```
todo : `const idToken = ` pas de destructuration du coup

Anne-Gaelle-S commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -4,15 +4,18 @@ const httpAgent = require('../../../infrastructure/http/http-agent');
     this.clientSecret = clientSecret;
     this.clientId = clientId;
     this.tokenUrl = tokenUrl;
+    this.authUrl = authUrl;
```
`authenticationUrl`

Anne-Gaelle-S commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -4,15 +4,18 @@ const httpAgent = require('../../../infrastructure/http/http-agent');
     this.clientId = clientId;
     this.tokenUrl = tokenUrl;
+    this.authUrl = authUrl;
+    this.urlParams = urlParams;
```
`authenticationUrlParameters`

Anne-Gaelle-S commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -54,6 +57,25 @@ class OidcAuthenticationService {
     });
   }
+
+  getAuthUrl({ redirectUri }) {
```
`getAuthenticationUrl`

Anne-Gaelle-S commented on api/tests/acceptance/application/pole-emploi/token-route-post_test.js:
```diff
@@ -87,7 +87,7 @@ describe('Acceptance | Route | pole emploi token', function () {
               given_name: 'John',
               family_name: 'Doe',
               nonce: 'nonce',
-              idIdentiteExterne: 'idIdentiteExterne',
```
à déplacer dans un autre commit

Anne-Gaelle-S commented on docs/security/poleEmploi/authenticate_open_id.part1.puml:
```diff
@@ -19,7 +19,7 @@ Pole_Emploi_Connect --> MonPix: Envoie de l'access_token
 MonPix -> Pole_Emploi_Connect: GET <API_PoleEmploi>/partenaire/peconnect-individu/v1/userinfo\nAuthorization: Bearer [Access token]
 note right: Demande d'information candidat Pôle Emploi
 Pole_Emploi_Connect --> MonPix: Envoie des informations du Candidat PE
-note right: email, family_name, given_name et idIdentiteExterne
```
merger avec le commit du sub

AndreiaPena commented on api/lib/domain/services/authentication/pole-emploi-authentication-service.js:
```diff
@@ -1,82 +1,7 @@
-    { key: 'response_type', value: 'code' },
-    {
-      key: 'scope',
-      value: `application_${clientId} api_peconnect-individuv1 openid profile serviceDigitauxExposition api_peconnect-servicesdigitauxv1`,
```
<img width="614" alt="Capture d’écran 2022-07-25 à 16 00 05" src="https://user-images.githubusercontent.com/58915422/180795543-d19b1254-7582-4652-8dad-8c97ce98f412.png">

d'après ce qui est obligatoire, on pourrait juste s'en tenir à 
`value: application_${clientId} api_peconnect-individuv1 openid`

Anne-Gaelle-S committed:
feat(api): add exchangeCodeForTokens in oidc authentication service

AndreiaPena committed:
feat(api): add getAuthUrl in oidc authentication service

AndreiaPena committed:
feat(api): add parameters in cnav and pe authentication services

AndreiaPena committed:
feat(api): use oidc authentication service new function in pole emploi and cnav controller

AndreiaPena committed:
recycle(api): remove unused functions

Anne-Gaelle-S committed:
feat(api): add createUserAccount in oidc authentication service

Anne-Gaelle-S committed:
feat(api): add createUserAccount in pole-emploi oidc authentication service

Anne-Gaelle-S committed:
feat(api): use createUserAccount from oidc service

AndreiaPena commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -19,6 +40,76 @@ class OidcAuthenticationService {
     );
   }
+
+  async exchangeCodeForTokens({ code, redirectUri }) {
```
On pourrait remettre l'ancien nom pour cette fonction, on ne récupère que l'idToken 

```suggestion
  async exchangeCodeForIdToken({ code, redirectUri }) {
```

AndreiaPena commented on api/tests/unit/domain/services/authentication/oidc-authentication-service_test.js:
```diff
@@ -31,4 +37,204 @@ describe('Unit | Domain | Services | oidc-authentication-service', function () {
+      expect(result).to.deep.equal(oidcAuthenticationSessionContent);
+    });
+
+    context('when tokens retrieval fails', function () {
```
```suggestion
    context('when id token retrieval fails', function () {
```

AndreiaPena commented on api/tests/unit/domain/services/authentication/oidc-authentication-service_test.js:
```diff
@@ -31,4 +37,204 @@ describe('Unit | Domain | Services | oidc-authentication-service', function () {
     });
   });
+
+  describe('#exchangeCodeForTokens', function () {
```
```suggestion
  describe('#exchangeCodeForIdToken', function () {
```

AndreiaPena committed:
recycle(api): remove unused functions

AndreiaPena committed:
recycle(api): return oidcAuthenticationService only in authenticationRegistry

yschoueri commented on api/lib/domain/services/authentication/pole-emploi-oidc-authentication-service.js:
```diff
@@ -1,15 +1,104 @@
+    });
+  }
+
+  // Overrided because we need accessToken, expiresIn and refreshToken to send results after a campaign
+  // Sending campaign results is specific for Pole Emploi
```
```suggestion
  // Overwritten because we need accessToken, expiresIn and refreshToken to send results after a campaign
  // Sending campaign results is specific to Pole Emploi
```

clemlatz commented on api/lib/domain/services/authentication/pole-emploi-oidc-authentication-service.js:
```diff
@@ -1,15 +1,52 @@
+  }
+
+  // Overrided because we need accessToken, expiresIn and refreshToken to send results after a campaign
+  // Sending campaign results is specific for Pole Emploi
```
Est-ce que c'est un problème de retourner les quatre aussi dans la méthode parente ? Je trouve dommage de dupliquer toute une méthode juste pour cette petite différence.

Si c'est un problème, je ferais plutôt le contraire : la méthode générique retourne les quatre (car il me semble que tout ça est dans le standard) et une méthode spécifique qui appelle la méthode générique puis filtre le retour pour ne renvoyer que l'idtoken :

```js
  async exchangeCodeForTokens({ code, redirectUri }) {
    const { idToken } = await super.exchangeCodeForTokens({ code, redirectUri });
    return { idToken };
  }
```

De cette manière, la grande partie du code en commun entre les deux méthodes n'est pas dupliqué.

AndreiaPena committed:
feat(api): add getRedirectLogoutUrl in poleEmploiOidcService

AndreiaPena committed:
recycle(api): remove unused getRedirectLogoutUrl function

AndreiaPena committed:
feat(api): add saveIdToken in poleEmploiOidcService

AndreiaPena committed:
feat(api): use oidc authentication service new function in pole emploi

AndreiaPena committed:
recycle(api): remove unused pole emploi service files

AndreiaPena committed:
recycle(api): remove unused redirectUri parameter in oidc controller

clemlatz commented on api/lib/application/pole-emploi/index.js:
```diff
@@ -76,7 +76,7 @@ exports.register = async function (server) {
       config: {
         auth: false,
-        handler: PoleEmploiController.getAuthUrl,
+        handler: PoleEmploiController.getAuthenticationUrl,
```
```suggestion
        handler: poleEmploiController.getAuthenticationUrl,
```

On dirait qu'on importe deux fois le même `poleEmploiController`, une fois en minuscule, une fois en majuscule…

AndreiaPena replied:
corrigé

AndreiaPena committed:
refactor(api): remove pole emploi controller import duplication

AndreiaPena commented on api/lib/domain/services/authentication/pole-emploi-oidc-authentication-service.js:
```diff
@@ -1,15 +1,105 @@
+
+    return {
+      userId: createdUserId,
+      idToken: sessionContent.idToken,
```
à voir

AndreiaPena commented:
TODO : supprimer le nextUrl dans le localStorage.

yschoueri commented on api/tests/unit/domain/services/authentication/oidc-authentication-service_test.js:
```diff
@@ -31,4 +37,216 @@ describe('Unit | Domain | Services | oidc-authentication-service', function () {
+  });
+
+  describe('#getAuthenticationUrl', function () {
+    it('should return auth url', async function () {
```
```suggestion
    it('should return authentication url', async function () {
```
